### PR TITLE
Fuzz testing for nomination pools

### PR DESCRIPTION
**Pull request by kianenigma**
This PR is itself pretty insubstantial, there is not real code change other than adding a fuzz-style test for nomination pools. 

Some notes about the approach: For now, I mainly did the fuzz-testing without using a proper fuzzer because it would allow for the `mock` to be reused. In the past we gated `mock.rs` behind some feature flag to make it available, and this time around I wanted to try the alternative approach. 

Having done this, I can say that the main advantage of this type of testing is seamless setup, and the downside is reproducibility of failing tests. 

For now, in order to have reproducible errors, you have to use a seeded randomness, and then re-use it. It requires a little bit of manual work, but it did allow me to solve every single test failure. 

It is probably still better to move the fuzz-test use a proper fuzzer as follow-up. Nonetheless, it must be done such that the mock file is reused (a good first issue for a new team member or external contributor).  

Some notes about the fuzzing outcome: As you see in the code, I randomly generate calls that may or may not be valid and bombard the pallet with these calls, and every so often I execute the entire set of sanity checks that are defined in the pallet. 

Moreover, I inject a single agent into a random pool that is manually tracking the expected reward amount (see `Agent` in the code for details). 

The tests generally get quite slow, particularly if you want to execute the sanity checks often. In the default setup, the state grows until 1000 pools and 10000 members, and with 1 million random submissions, this is the final outcome: 

```
Aug 10 16:16:23.564  INFO runtime::nomination-pools: [1] 🏊‍♂️ running sanity checks at 1010000
Aug 10 16:16:23.640  INFO runtime::nomination-pools: [1] 🏊‍♂️ iteration 1010000, 1000 pools, 10000 members, 325223 ok 684777 err, events = [("Event::Created", 1000), ("Event::Bonded", 72300), ("Event::Unbonded", 41458), ("Event::PaidOut", 164750), ("Event::Withdrawn", 21622), ("Event::MemberRemoved", 1)]
```

All in all, this was super useful and I feel way more confident about nomination pools now. 

- [ ] follow-up Issue to migrate this to a proper fuzzer.

### CODE REVIEW

File: frame/support/src/lib.rs [hunk 1]
```diff
@@ -428,6 +428,13 @@ macro_rules! parameter_types_impl_thread_local {
 					pub fn set(t: $type) {
 						[<$name:snake:upper>].with(|v| *v.borrow_mut() = t);
 					}
+
+					/// Mutate the internal value in place.
+					pub fn mutate<F: FnOnce(&mut $type) -> ()>(mutate: F) {
```
Comment (kianenigma) [inline frame/support/src/lib.rs:433]: This is a unrelated change -- small enough to sneak in here, but I can also extract.

File: primitives/runtime/src/lib.rs [hunk 2]
```diff
@@ -825,7 +827,7 @@ pub fn verify_encoded_lazy<V: Verify, T: codec::Encode>(
 macro_rules! assert_eq_error_rate {
 	($x:expr, $y:expr, $error:expr $(,)?) => {
 		assert!(
-			($x) >= (($y) - ($error)) && ($x) <= (($y) + ($error)),
+			($x >= $crate::Saturating::saturating_sub($y, $error)) && ($x <= $crate::Saturating::saturating_add($y, $error)),
```
Comment (kianenigma) [inline primitives/runtime/src/lib.rs:830]: This is a unrelated change -- small enough to sneak in here, but I can also extract.

**Review by sam0x17 — APPROVED** [review]
Really cool to see some fuzzing. Looks good to me!

**Comment by ggwpez** [comment]
> Some notes about the fuzzing outcome: As you see in the code, I randomly generate calls that may or may not be valid and bombard the pallet with these calls, and every so often I execute the entire set of sanity checks that are defined in the pallet.  

We could create a tool which generates a set of calls from the metadata of a pallet. Could even be done in PolkadotJS, as general fuzzy testing framework.

File: frame/nomination-pools/src/tests.rs [hunk 3]
```diff
@@ -4768,3 +4980,319 @@ mod reward_counter_precision {
...
+		assert!(Balances::free_balance(account) >= target_amount);
+	}
+
+	fn random_existing_pool<R: Rng>(mut rng: &mut R) -> Option<PoolId> {
+		BondedPools::<T>::iter_keys().collect::<Vec<_>>().choose(&mut rng).map(|x| *x)
+	}
+
+	fn random_call<R: Rng>(mut rng: &mut R) -> (crate::pallet::Call<T>, Origin) {
+		let op = rng.gen::<u8>();
+		match op % 8 {
```
Comment (ggwpez) [inline frame/nomination-pools/src/tests.rs:5041]:
```suggestion
		let op = rng.gen::<usize>();
		let mut max_op_index =
			<crate::pallet::Call<T> as frame_support::dispatch::GetCallName>::get_call_names()
				.len();
		// Exclude set_state, set_metadata, set_configs, update_roles and chill.
		max_op_index -= 5;

		match op % max_op_index {
```
Could be overkill but like this you wont forget to add future extrinsics.

File: frame/nomination-pools/src/tests.rs [hunk 4]
```diff
@@ -4768,3 +4980,319 @@ mod reward_counter_precision {
...
+	}
+
+	#[test]
+	fn fuzz_test() {
+		let mut reward_agent = RewardAgent::new(42);
+		sp_tracing::try_init_simple();
+		// NOTE: use this to get predictable (non)randomness:
+		// use::{rngs::SmallRng, SeedableRng};
+		// let mut rng = SmallRng::from_seed([0u8; 32]);
+		let mut rng = thread_rng();
```
Comment (ggwpez) [inline frame/nomination-pools/src/tests.rs:5167]: Good follow up:

From experience I can tell you that it is a good idea to create a random seed, print it, and add the possibility to overide it via env variable.  
Then you always have a different set of tests, but if it fails, you can reproduce by using the printed seed.

File: frame/nomination-pools/src/tests.rs [hunk 5]
```diff
@@ -4768,3 +4980,319 @@ mod reward_counter_precision {
...
+		let mut ext = sp_io::TestExternalities::new_empty();
+		// NOTE: sadly events don't fulfill the requirements of hashmap or btreemap.
+		let mut events_histogram = Vec::<(PoolsEvents<T>, u32)>::default();
```
Comment (ggwpez) [inline frame/nomination-pools/src/tests.rs:5170]: Is it not enough to use the event index, or do you also want the values?

File: frame/nomination-pools/src/tests.rs [hunk 6]
```diff
@@ -4768,3 +4980,319 @@ mod reward_counter_precision {
...
+				}
+				// and possibly roughly every 4 era, trigger payout for the agent. Doing this more
+				// frequent is also harmless.
+				if rng.gen_range(0..(4 * ERA)) == 0 {
+					reward_agent.claim_payout();
+				}
+
+				// execute sanity checks at a fixed interval, possibly on every block.
+				if iteration %
+					(std::env::var("SANITY").ok().and_then(|x| x.parse::<u64>().ok()))
```
Comment (ggwpez) [inline frame/nomination-pools/src/tests.rs:5222]:
```suggestion
					(std::env::var("SANITY_CHECK_INTERVAL").ok().and_then(|x| x.parse::<u64>().ok()))
```

File: frame/nomination-pools/src/tests.rs [hunk 7]
```diff
@@ -4768,3 +4980,319 @@ mod reward_counter_precision {
...
+							.find(|(x, _)| std::mem::discriminant(x) == std::mem::discriminant(&e))
+						{
+							*c += 1;
+						} else {
+							events_histogram.push((e, 1))
+						}
+					});
+				System::reset_events();
+
+				if iteration % ERA == 0 {
```
Comment (ggwpez) [inline frame/nomination-pools/src/tests.rs:5248]: Some smol doc please.

**Review by ggwpez — APPROVED** [review]

File: frame/nomination-pools/src/tests.rs (hunk 4 above)
Comment (kianenigma) [reply frame/nomination-pools/src/tests.rs:5167]: excellent idea, I actually have this in my code as a commented code that: If you want reproducible output, do this and that, but using a seed and env var makes a whole lot more sense.

File: frame/nomination-pools/src/tests.rs (hunk 3 above)
Comment (kianenigma) [reply frame/nomination-pools/src/tests.rs:5041]: bueno

**Pull request merged**